lib/ansible/plugins/connection/paramiko_ssh.py

```diff
@@ -279,14 +300,21 @@ def missing_host_key(self, client, hostname, key):
 
             self.connection.connection_lock()
 
-            old_stdin = sys.stdin
-            sys.stdin = self._new_stdin
+            if self._new_stdin is not None:
+                old_stdin = sys.stdin
+                sys.stdin = self._new_stdin
+
+                # clear out any premature input on sys.stdin
+                tcflush(sys.stdin, TCIFLUSH)
 
-            # clear out any premature input on sys.stdin
-            tcflush(sys.stdin, TCIFLUSH)
+                inp = input(AUTHENTICITY_MSG % (hostname, ktype, fingerprint))
+                sys.stdin = old_stdin
+            else:
```

Comment:
The original code was already doing this "wrong" IMO anyway- a blind asynchronous call to Python `input` is fraught with peril in a world where `display` expects roughly exclusive control of stdout/stderr. Letting `display` lose the context that we're doing a prompt by separating the output from the input with two calls and a relative eternity for things to change from underneath us between multiple threads/processes and a bunch of IPC isn't going to improve the situation :laughing: .

I've actually got a working way to just fully remote the entire display object from the controller into the worker (and vice versa for any arbitrary objects) without an external `Manager`, but I don't think we're ready to do that just yet.

With a fairly small change, we can just remote a call to the existing native `display.prompt` on the controller side, similarly to what we're doing with `display.display` now. You'd just do like `display.display` did and delegate to a worker-local queue-backed implementation of `display.prompt` that uses the dataclass you've already created (maybe with some slight tweaks) and sends it to the controller over the queue (including the message to display) and blocks on the queue for its response and returns/raises appropriately. While the worker is waiting for its result, once the controller has dequeued the PromptSend, it just calls the real `display.prompt` with the values you passed over, then sends the response back through the worker's queue. This mechanism doesn't scale that well, but for how infrequently these kinds of prompts happen, it's fine, and the whole thing can go away once we've got arbitrary bidrectional object remoting anyway. This way, the plugin code can still just directly call the appropriate methods on `display`, which is much clearer anyway, and noone's the wiser when we change how that's actually occurring, whether it's over a queue, a remoted `Display` object, or a carrier pigeon.

Remember too, that even though there's a lock to prevent >1 connection from doing this concurrently, there's potentially a *bunch* of unrelated stuff happening concurrently that's talking to `Display` while we're waiting for someone to answer these prompts, so its state can change in all sorts of problematic ways. If we really want this to be robust, we'll also need to synchronize or queue writes to `display` on the controller, such that a pending prompt will prevent anything else from writing to stdout. That part might be "bonus points" for this PR, since from what I can see, we don't do much (if anything) to prevent that right now anyway, so this change doesn't make that any worse than it already is. If we did that, though, we could actually remove the `connection_lock/unlock` stuff entirely, since the plugin will just naturally block until the controller gets back to it, and the controller's `Display` class is the one arbitrating all the actual concurrency, state of stdout, and listening on stdin for one prompt at a time.

Reply:
I think I've addressed this in the latest non-documentation commit. I renamed the method to `prompt_until` since it doesn't just read now, and pass a message as the first argument. I'm returning the value from the queue after sending the prompt, so plugins can just use the display method.

I thought I would be able to refactor `prompt_until` so paramiko_ssh could still call `input()` via the queue (and maybe one of the corner cases for `pause` - if there is no timeout and we're waiting for enter), but I didn't realize KeyboardInterrupt (which `input()` can raise), needs to be handled in the main thread (as opposed the daemon results thread). Here's a simple example https://gist.github.com/s-hertel/ebd57439ee74d0e40028ccb5fd6a2d78 to visualize the issue + possible fix. If the exception handling needs to occur in the main thread, then the interrupt/done logic would also need to be moved there as well, so I left a comment about why `self.prompt` isn't used when it seems like it should be for the default use-case.

I'm not sure I've addressed your last point, but I have now manually tested the paramiko_ssh prompt against a dozen hosts, and it is working with free/linear/host_pinned without the connection_lock/unlock, so I removed it. (The display lock *is* used in `prompt_until` since I setup/cleanup for each prompt.)